ocsp/responder/multi_source_test.go

```diff
@@ -30,6 +32,40 @@ func (src *failSource) Response(context.Context, *ocsp.Request) (*Response, erro
 	return nil, errors.New("failure")
 }
 
+// timeoutSource is a Source that will not return until its chan is closed.
+type timeoutSource struct {
+	ch <-chan struct{}
+}
+
+func (src *timeoutSource) Response(context.Context, *ocsp.Request) (*Response, error) {
+	<-src.ch
+	return nil, errors.New("failure")
+}
+
+func TestSecondaryTimeout(t *testing.T) {
+	ch := make(chan struct{})
+	src, err := NewMultiSource(&succeedSource{}, &timeoutSource{ch: ch}, metrics.NoopRegisterer, blog.NewMock())
+	test.AssertNotError(t, err, "failed to create multiSource")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+
+	starting_goroutines := runtime.NumGoroutine()
+
+	for i := 0; i < 1000; i++ {
+		_, err = src.Response(ctx, &ocsp.Request{})
+		test.AssertNotError(t, err, "unexpected error")
+	}
+
+	close(ch)
+	// Wait for the goroutines to exit
+	time.Sleep(40 * time.Millisecond)
+	goroutine_diff := runtime.NumGoroutine() - starting_goroutines
+	if goroutine_diff > 0 {
```

Comment:
Somehow this _feels_ fragile. For example, if this test were ever run in parallel with other tests which are spinning up and down their own goroutines, this would likely fail. I know that the go test runtime guarantees that only tests marked with `t.Parallel` will be run in parallel, but it still feels like this is relying on other assumptions. What if a garbage collector goroutine spins up in the background? Maybe I just don't have a deep enough understanding of the go runtime to be confident this won't flake.